packages/vite/src/node/server/middlewares/indexHtml.ts

```diff
@@ -41,6 +41,31 @@ function getHtmlFilename(url: string, server: ViteDevServer) {
 }
 
 const startsWithSingleSlashRE = /^\/(?!\/)/
+const processUrlNode = (
+  node: AttributeNode,
+  s: MagicString,
+  config: ResolvedConfig
+) => {
+  const url = node.value?.content || ''
+  if (startsWithSingleSlashRE.test(url)) {
+    // prefix with base
+    s.overwrite(
+      node.value!.loc.start.offset,
+      node.value!.loc.end.offset,
+      `"${config.base + url.slice(1)}"`
+    )
+  }
+  // #3230 if some request url (localhost:3000/a/b) return to fallback html, the relative assets
+  // path will add `/a/` prefix, it will caused 404.
+  // here add prefix `/@fs/` for the relative assets path.
+  if (url.startsWith('.')) {
```

Comment:
Should we add an `else` here so it is clear that only one branch would be called?

Reply:
Agreed. Other than that LGTM